instrumentation/github.com/Shopify/sarama/otelsarama/producer.go

```diff
@@ -189,41 +173,73 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
 
 				p.Input() <- msg
 				if saramaConfig.Producer.Return.Successes {
+					mtx.Lock()
 					producerMessageContexts[msg.Metadata] = mc
+					mtx.Unlock()
 				} else {
 					// If returning successes isn't enabled, we just finish the
 					// span right away because there's no way to know when it will
 					// be done.
-					span.End()
-				}
-			case msg, ok := <-p.Successes():
-				if !ok {
-					// producer was closed, so exit
-					return
+					mc.span.End()
 				}
-				key := msg.Metadata
-				if mc, ok := producerMessageContexts[key]; ok {
-					delete(producerMessageContexts, key)
-					finishProducerSpan(mc.span, msg.Partition, msg.Offset, nil)
+			}
+		}
+	}()
 
-					// Restore message metadata
-					msg.Metadata = mc.metadataBackup
-				}
-				wrapped.successes <- msg
-			case err, ok := <-p.Errors():
-				if !ok {
-					// producer was closed
-					return
-				}
-				key := err.Msg.Metadata
-				if mc, ok := producerMessageContexts[key]; ok {
-					delete(producerMessageContexts, key)
-					finishProducerSpan(mc.span, err.Msg.Partition, err.Msg.Offset, err.Err)
-				}
-				wrapped.errors <- err
+	// Spawn spans cleanup goroutine.
+	// Sarama will consume all the successes and errors by itself while closing,
+	// so we need to end these these spans ourseleves.
+	cleanupCh := make(chan struct{}, 2)
```

Comment:
I would use a `sync.WaitGroup` for this instead.  `wg.Add(1)`/`wg.Done()` in the cleanup goroutines and `wg.Wait()` before doing the work here.

Reply:
I tried it initially. It produces a data race (sic!). It looks like Add and Wait have to be used in the same goroutine. I will just try it again. Maybe I made some silly error